decisions/DR009-Batch-method-result-types.md

```diff
@@ -97,7 +97,176 @@ performant c++ to c++ path.
+has the form
+
+```c++
+using SuccessCallback = std::function<void(std::size_t, TraitsDataPtr)>;
+using ErrorCallback = std::function<void(std::size_t, ErrorCodeAndMessage)>;
+
+void resolve(const std::vector<EntityRef>& entityRefs,
+             const std::unordered_set<std::string>& traitSet,
+             const ContextPtr& context,
+             const managerAPI::HostSessionPtr& hostSession
+             const SuccessCallback& successCallback,
+             const ErrorCallback& errorCallback);
+```
+where the first `std::size_t` argument to a callback is the relevant
+index in the input array of entity references.
+
+## Appendix 2 - Performance
+
+To test the performance of the two approaches, the perftest binary was
+updated to evaluate the above sketches in roughly representative
+scenarios. The code can be found in the repository under
+`resources/perftest/perftest.cpp` (future performance tests may
+overwrite this, but the code as used for the following tests will remain
+accessible via the git history).
+
+When deciding on the performance impact of one design over another, we
+had several concerns
+
+1. Our main concern: is it cheaper to call a callback repeatedly, once
+   for each result, or is it cheaper to return an array of results and
+   iterate over it?
+2. Assuming the user _wants_ an array of results, what is the cost of
+   using a callback design anyway, and have the user construct their own
+   array?
+3. Is there a performance benefit to using raw function pointers rather
+   than `std::function`?
+4. What is the performance impact of the small-size optimization in
+   common `std::function` implementations?
+
+### Test setup
+
+The following results come from a build using with GCC 9.3 on Ubuntu
+20.04, compiled in a Release (`O3`) configuration.
+
+Each set of cases is executed over 5000 iterations (epochs), with a
+warm-up of 1000 epochs to alleviate CPU scaling artifacts, etc. In
+addition, the whole test is executed 5 times (5 runs of the perftest
+executable) and the summary results with the lowest standard deviation
+taken, on the assumption that this represents the results that are least
+skewed by external factors.
+
+A database (simple map of entity reference to URL) of size 10000 is
+generated to draw from. Input entity references are randomly chosen from
+this database, with the probability of an erroneous entity reference set
+to 0.1. Timings are nanosecond precision.
+
+Since we're concerned with the relative performance, the key output of
+the perftest is the ratio of timings between two solutions. The mean and
+standard deviation of these timing ratios are reported in the tables
+below. We round to 3 significant figures - the resolution of the fastest
+results are 3 significant figures (hundreds of nanoseconds).
+
+As shorthand for the different test cases, we define
+* _vector_ as the case that returns an array of results.
+* _callback_ as the case that executes a `std::function` callback on
```

Comment:
[fix] Could we really spell it out and have `(Option 1)` and `(Option 2)` on these two?

Reply:
Hopefully clarified in 1f0f5ae5